### PR TITLE
rgw: rework the Swift TempAUTH's implementation

### PR DESCRIPTION
**Pull request by rzarzynski**

### CODE REVIEW

File: src/rgw/rgw_common.h [hunk 1]
```diff
@@ -2079,6 +2079,27 @@ static inline void buf_to_hex(const unsigned char* const buf,
...
+/* A variant of buf_to_hex dedicated to handling the run-time bufferlist
+ * which can't be hexed in compile-time. It's optimized to avoid extra
+ * linearisation of the incoming buffers. */
+static inline std::string buf_to_hex(const ceph::bufferlist& bl)
+{
+  std::string hexed;
+  hexed.resize(bl.length() * 2);
+
+  size_t consumed = 0;
+  for (const auto buf : bl.buffers()) {
```
Comment (cbodley) [inline src/rgw/rgw_common.h]: copying the `buffer::ptr` here induces an atomic increment and decrement of `buffer::raw::nref`, so it's better to take these by reference

Comment (rzarzynski) [reply src/rgw/rgw_common.h]: Thanks for pointing this out, @cbodley! Reworked.

File: src/rgw/rgw_swift_auth.cc [hunk 2]
```diff
@@ -203,13 +201,17 @@ class TempURLEngine::PrefixableSignatureHelper
   const boost::optional<const std::string&> prefix;
 
 public:
-  PrefixableSignatureHelper(const std::string& decoded_uri,
+  PrefixableSignatureHelper(const std::string& _decoded_uri,
```
Comment (amitkumar50) [inline src/rgw/rgw_swift_auth.cc]: @rzarzynski Is there an convention to prefix underscore in string name. 
Though I am not aware. Or need for rados gateway. Sorry for that

File: src/rgw/rgw_swift_auth.cc [hunk 3]
```diff
@@ -418,21 +420,20 @@ ExternalTokenEngine::authenticate(const std::string& token,
   return result_t::grant(std::move(apl));
 }
 
-static int build_token(const string& swift_user,
-                       const string& key,
-                       const uint64_t nonce,
-                       const utime_t& expiration,
-                       bufferlist& bl)
+static ceph::bufferlist build_token(CephContext* const cct,
+                                    const std::string& swift_user,
```
Comment (amitkumar50) [inline src/rgw/rgw_swift_auth.cc]: tabs missed

File: src/rgw/rgw_swift_auth.cc [hunk 4]
```diff
@@ -418,21 +420,20 @@ ExternalTokenEngine::authenticate(const std::string& token,
...
   ::encode(swift_user, bl);
   ::encode(nonce, bl);
   ::encode(expiration, bl);
 
   bufferptr p(CEPH_CRYPTO_HMACSHA1_DIGESTSIZE);
 
-  char buf[bl.length() * 2 + 1];
-  buf_to_hex((const unsigned char *)bl.c_str(), bl.length(), buf);
-  dout(20) << "build_token token=" << buf << dendl;
+  ldout(cct, 20) << "build_token token=" << buf_to_hex(bl) << dendl;
```
Comment (amitkumar50) [inline src/rgw/rgw_swift_auth.cc]: I believe this should be enough:
`ldout(cct, 20) << "build_token token=" << buf_to_hex(bl) << dendl;`
`token` seems redundant here

File: src/rgw/rgw_swift_auth.cc [hunk 5]
```diff
@@ -636,62 +618,110 @@ void RGW_SWIFT_Auth_Get::execute()
...
-  if (!g_conf->rgw_swift_tenant_name.empty()) {
-    tenant_path = "/AUTH_";
-    tenant_path.append(g_conf->rgw_swift_tenant_name);
-  } else if (g_conf->rgw_swift_account_in_url) {
-    tenant_path = "/AUTH_";
-    tenant_path.append(info.user_id.to_str());
+  const auto siter = info.swift_keys.find(user_str);
+  if (siter == std::end(info.swift_keys)) {
+    ldout(s->cct, 20) << "TempAUTH: can't find swift key for uid="
+                      << user_str << dendl;
```
Comment (amitkumar50) [inline src/rgw/rgw_swift_auth.cc]: tab missed

File: src/rgw/rgw_common.h [hunk 6]
```diff
@@ -2079,6 +2079,29 @@ static inline void buf_to_hex(const unsigned char* const buf,
...
+ * The procedure is optimized to avoid extra linearisation of incoming
+ * buffers. */
+static inline std::string buf_to_hex(const ceph::bufferlist& bl)
+{
+  std::string hexed;
+  hexed.resize(bl.length() * 2);
+
+  size_t consumed = 0;
+  for (const auto& buf : bl.buffers()) {
+    /* NOTE(rzarzynski): yes, we're writing directly to underlying buffer
```
Comment (amitkumar50) [inline src/rgw/rgw_common.h]: Please remember to remove name `rzarzynski`.
May be later on PR would be raised for removal.

**Comment by mattbenjamin** [comment]
@mdw-at-linuxbox could you look this over, see if we can still absorb it?

**Pull request closed without merging**